### ISSUE TITLE
Caching database from network filesystem

### ISSUE THREAD

**Issue by LorenzoFioroni** [keeweb/keeweb#1789]
Hi,
I'm experiencing an issue caching a password database stored in my NAS mounted as a nfs.

**Describe the bug**
When I'm at home and the NAS is mounted as nfs I can successfully open and use my password database. If I try to open the same file from the recently used files list when I'm not at home and the NAS is unmounted, though, I receive a "File not found" message. I've read in issue #593 that it isn't the expected behaviour. I've found in the settings the option "Don't cache loaded files in offline storage" which is toggled off.

**Expected behavior**
I would expect a cached copy of the db to be opened.

**Environment**
KeeWeb v1.17.6 (8efee7d, 2021-04-09)
Environment: electron v12.0.0
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) KeeWeb/1.17.6 Chrome/89.0.4389.69 Electron/12.0.0 Safari/537.36

**Logs**
```
[INFO ] 2021-04-19T09:17:01.027Z [open:portkey] File open request
[INFO ] 2021-04-19T09:17:01.027Z [open:portkey] Open file from storage file
[INFO ] 2021-04-19T09:17:01.027Z [open:portkey] Stat file
[DEBUG] 2021-04-19T09:17:01.028Z [storage-file] Stat /home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx
[DEBUG] 2021-04-19T09:17:01.072Z [storage-file] Stat done /home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx 45ms
[INFO ] 2021-04-19T09:17:01.073Z [open:portkey] Open file from storage (1618760653739, local 1618760653739)
[INFO ] 2021-04-19T09:17:01.073Z [open:portkey] Load from storage
[DEBUG] 2021-04-19T09:17:01.073Z [storage-file] Load /home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx
[DEBUG] 2021-04-19T09:17:01.258Z [storage-file] Loaded /home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx 1618760653739 185ms
[INFO ] 2021-04-19T09:17:01.258Z [open:portkey] Open file from content loaded from storage
[DEBUG] 2021-04-19T09:17:01.264Z [argon2] Using native argon2
[DEBUG] 2021-04-19T09:17:01.264Z [native-module-connector] Starting native module host
[DEBUG] 2021-04-19T09:17:01.282Z [argon2] Argon2 hash calculated 18ms
[DEBUG] 2021-04-19T09:17:01.282Z [argon2] Hash computed 18ms
[INFO ] 2021-04-19T09:17:01.778Z [file] Opened file portkey: 518ms, P=2&V=19&I=20&M=65536, 3077 kB
[DEBUG] 2021-04-19T09:17:01.778Z [app] Add last open file 3fe1cd3a-33bb-ed13-aee5-a8f9181f2deb portkey file /home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx 1618760653739
[DEBUG] 2021-04-19T09:17:01.921Z [storage-file] Watch dir /home/lorfioroni/Fritz-nas/Elements/Lorenzo
[DEBUG] 2021-04-19T09:17:04.473Z [storage-file] Stop watch dir /home/lorfioroni/Fritz-nas/Elements/Lorenzo
[INFO ] 2021-04-19T09:17:13.555Z [open:portkey] File open request
[INFO ] 2021-04-19T09:17:13.555Z [open:portkey] Open file from storage file
[INFO ] 2021-04-19T09:17:13.555Z [open:portkey] Stat file
[DEBUG] 2021-04-19T09:17:13.555Z [storage-file] Stat /home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx
[ERROR] 2021-04-19T09:17:13.563Z [storage-file] Error stat local file /home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx Error: ENOENT: no such file or directory, stat '/home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx'
[INFO ] 2021-04-19T09:17:13.563Z [open:portkey] Stat error Error: ENOENT: no such file or directory, stat '/home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx'
[ERROR] 2021-04-19T09:17:13.564Z [open-view] Error opening file Error: ENOENT: no such file or directory, stat '/home/lorfioroni/Fritz-nas/Elements/Lorenzo/portkey.kdbx'
```
Thanks,

Lorenzo

**Comment by antelle** [comment]
A mounted folder is essentially just a filesystem, so we don't cache from there. Probably there should be a setting "cache files from disk" or so.

**Comment by LorenzoFioroni** [comment]
So the issue #593 is not fixed or it refers to something else? Anyway it would be great, thanks!

**Comment by antelle** [comment]
Interesting, not sure why I closed that issue 🤔 . Saving local files to cache was disabled in [v1.2.0](https://github.com/keeweb/keeweb/blob/master/release-notes.md#v120-2016-07-11), since then it was never enabled or added as an option. Now we also have an option to skip cache for files from cloud, perhaps this should be somewhere next to it.

**Comment by LorenzoFioroni** [comment]
I think it would be an useful option. Domestic NAS are not so uncommon and saving the password file there seems to me as the natural solution.
I understand that it is not an urgent feature, but let me know if I can be of any help.
Thanks
Lorenzo

**Comment by antelle** [comment]
You can try reverting c7325076cac49c94506626ba9ae26817ca81d8a4 and wrapping in an option, if you would like to give it a try. I assume, the revert won't go easy because the file content has diverged.

**Comment by LorenzoFioroni** [comment]
Hi Antelle, 
I'm trying to implement what we said but I actually have 2 doubts regarding the software design:

- Now there is an option with label _"Don't cache loaded files in offline storage"_. Does it refer to the cloud only or is it in general? Because in the latter case the _"Cache files from filesystem"_ option should be conditioned to the other, in my opinion.

- When I try to open a previously cached file from Google Drive and I'm offline, I can see from the console that the stat process exits with "Network error". If the same procedure is carried out opening the file from the local filesystem though, if the file is not present (i.e. filesystem not mounted), it gives an "ENOENT error" and tries to create the file. Now, since the fs is not mounted, the creation process fails. Do you think this acceptable or it should be implemented a second check after the ENOENT to know whether is it the file which does not exists (and therefore should be created) or the folder which is not mounted?

Thanks a lot and sorry for the questions but it is the very first time that I try to contribute to such a big project :wink:

**Comment by antelle** [comment]
Yes, having two options would be confusing, that's why I suggested to offer this option like this:

Cache files in the offline storage:
- [x] loaded from cloud
- [ ] local files

> it gives an "ENOENT error" and tries to create the file

Hm that's strange, it shouldn't create files, what leads to file creation? If you just open a file in KeeWeb, then lock KeeWeb and delete the file - it doesn't create a new file for me.

**Comment by LorenzoFioroni** [comment]
I'm sorry, that's true, it doesn't create anything. I edited the code and since my fs was not mounted it entered in
```javascript
if (err.notFound) {
  logger.info('File does not exist in storage, creating');
  saveToCacheAndStorage();
}
```
my bad. 

Regarding the options, it definitely is clearer in this way. Thanks!

**Comment by antelle** [comment]
That's another piece of code, the sync code is somewhat complicated because it handles many different scenarios.
